### ISSUE TITLE
Make the API for watching of spent outpoints on Bitcoin misuse resistant

### ISSUE THREAD

**Issue by thomaseizinger** [comit-network/comit-rs#2270]
## Problem

The current API for watching spent outpoints looks like this:

https://github.com/comit-network/comit-rs/blob/0cbf97d91edf318707b17c62c8b9c7fe7fa739c1/cnd/src/btsieve/bitcoin/mod.rs#L39-L44

Notice the `unlock_script` parameter?

This allows you to match on any part of the witness stack to determine, how a specific outpoint is spent. We use this to differentiate between redeemed and refunded HTLCs.

This is problematic:

1. With the planned move to miniscript in https://github.com/comit-network/blockchain-contracts/issues/23, we can no longer pass `1` or `0` here because the script generated by miniscript does not use a single byte on the witness stack to differentiate between redeem and refund. Instead if tries to match the signature directly and checks, whether it was succesful or not.

2. The API is depending on implementation details of our HTLCs which are defined in a completely different crate. To use the API correctly, you need to know, whether the redeem path is triggered with a 1 or 0. This creates an extremely high risk to be misused. Potentially, this is even a security problem if our script is malleable and an attacker can somehow sneak a 1 or a 0 in there. The bottom line is: Only relying on a `0` or a `1` is very bad.

This API is like that only for historical reasons: It used to be part of the btsieve binary, which didn't know about HTLCs. So we had to match on something that we certainly knew was going to be part of the stack.

## Goal

With btsieve being fully-integrated into cnd, we can now design the API in exactly the way we need it to be and make it a lot safer to use.

The goal is to make this API misuse resistant, that means: whatever we pass in there, it should be unambiguous about whether we want to match the refund or the redeem path of the script.

## Recommendation

Keep in mind that with miniscript, we don't have control over specific details of the script.

Here are some ideas how we could go about this:

1. Pass in a public key that we want to appear on the stack.

It should be safe to assume that public keys only appear on the stack if we want to verify a signature with them. Hence, if the refunder's public key appears on the stack, it is very likely that the HTLC was refunded.

**Comment by thomaseizinger** [comment]
Labelled this as technical-debt because it is only like that for historical reasons. The circumstances under which this design was created changed completely.

**Comment by bonomat** [comment]
@thomaseizinger : shall we block this on the mini-script issue: #1266 to prevent a redesign of the API?

**Comment by thomaseizinger** [comment]
@bonomat I hestiated but opted not to because 

a) the only recorded miniscript issue is in blockchain-contracts and hence can be done completely independently of this
b) I think it is a good idea to do this as part of the task of integrating said work into comit-rs. While it may be possible to do without if we carefully examine the witness stack that miniscript creates, it seems like the right point in time to fix this API.

That being said, nothing prevents us from fixing it earlier if we want to :)
Doing the miniscript ticket in blockchain-contracts earlier will give us a concrete idea what the witness stack looks like and hence more certainty that the proposed solution will work for that as well.

So, long story short: it may be smart to move blockchain-contracts to miniscript first, but it is not a hard dependency.
